File: LeetCode/1415/cpp/lc1415.cpp
#include <iostream>
#include <queue>
#include <unordered_map>

class Solution {
public:
    std::string getHappyString(int n, int k) {
        const std::unordered_map<char, std::string> letterToNext = {
            {'a', "bc"},
            {'b', "ac"},
            {'c', "ab"}
        };

        std::queue<std::string> q{{"a", "b", "c"}};

        while (q.front().size() != n) {
            std::string base = q.front();
            q.pop();
            for (char postfix: letterToNext.at(base.back())) { // note: .at() fixes "no viable operator[]".
                q.push(base + postfix);
            }
        }
        if (q.size() < k) {
            return "";
        }
        for (int i = 1; i < k; i++) { // note: starting from 1, so that (k-1) entries are removed. Then k-th is the one on top.
            q.pop();
        }
        return q.front();

    }
};

int main() {
    struct Testcase {
        int n;
        int k;
        std::string expected;
    };
    Testcase testcases[] = {
        {1, 3, "c"},
        {1, 4, ""},
        {3, 9, "cab"},
    };
    Solution s{};
    int numGood = 0, numBad = 0;
    for (const Testcase& tc: testcases) {
        auto actual = s.getHappyString(tc.n, tc.k);
        if (actual != tc.expected) {
            std::cout << "Testcase (n=" << tc.n << ", k=" << tc.k << ") failed. Got: " << actual
                << ", want: " << tc.expected << "\n";
            ++numBad;
        } else {
            ++numGood;
        }
    }
    std::cout << (numBad == 0 ? "[OK]" : "[FAIL]") << " "
        << numGood << "/" << (numBad + numGood) << " testcases passed successfully.\n";
}

File: LeetCode/1415/golang/lc1415.go
package lc1415

import "fmt"

type StringQueue []string

func (q *StringQueue) Enqueue(val string) {
	*q = append(*q, val)
}

func (q *StringQueue) Dequeue() (string, error) {
	if len(*q) == 0 {
		return "error", fmt.Errorf("Dequeue() from an empty Queue.")
	}
	val := (*q)[0]
	*q = (*q)[1:]
	return val, nil
}

// intentionally no bounds check; use only on confirmed Len() > 0.
func (q *StringQueue) UnsafeFront() string {
	return (*q)[0]
}

func (q *StringQueue) Len() int {
	return len(*q)
}

// Runtime complexity: O(3^n)
// Auxiliary space: O(3^n) - for the queued items
// (there are exactly 3^n happy strings of length n)
// Subjective level: medium
// Solved on: 2026-03-14
func getHappyString(n int, k int) string {
	letterToNext := map[byte]string{
		'a': "bc",
		'b': "ac",
		'c': "ab",
	}
	q := StringQueue{"a", "b", "c"}
	for len(q.UnsafeFront()) < n {
		base, _ := q.Dequeue()
		lastBaseChar := base[len(base)-1]
		for _, ch := range letterToNext[lastBaseChar] {
			q.Enqueue(base + string(ch))
		}
	}

	if q.Len() < k {
		return ""
	}

	// note: starting from 1, so that (k-1) entries are removed.
	// Then the k-th entry is the one on top.
	for i := 1; i < k; i++ {
		_, _ = q.Dequeue()
	}
	return q.UnsafeFront()
}
